### PR TITLE
Permissions to master

### PR DESCRIPTION
**Pull request by vu3rdd**
Permissions code, which is already reviewed and merged into `bryan/upstream-wip` rebased to `master`.
---

## Code Review Checklist (to be filled out by reviewer)

- [ ] Description accurately reflects what changes are being made.
- [ ] Description explains why the changes are being made (or references an issue containing one).
- [ ] The PR appropriately sized.
- [ ] New code has enough tests. 
- [ ] New code has enough documentation to answer "how do I use it?" and "what does it do?".
- [ ] Existing documentation is up-to-date, if impacted.

### CODE REVIEW

**Comment by vu3rdd** [comment]
I see a CI failure and can't seem to reproduce it on my local setup. :-(

**Comment by vu3rdd** [comment]
I think the CI flakiness was with the `i386` test that exercises the hashcash based permissions which had a large number of "required bits" for the hashcash. For the test I reduced it from 20 bits to 10 bits and the CI flakiness seem to have vanished. I tried re-running tests and didn't see any issues.

File: wasm/client.go [hunk 1]
```diff
@@ -205,15 +206,22 @@ func Client_SendFile(_ js.Value, args []js.Value) interface{} {
 			return
 		}
 
+		go func() {
+			<-ctx.Done()
+			fmt.Printf("client.SendFile: received cancel signal\n")
```
Comment (bryanchriswhite) [inline wasm/client.go:211]:
```suggestion
```

File: wasm/client.go [hunk 2]
```diff
@@ -236,6 +244,7 @@ func Client_SendFile(_ js.Value, args []js.Value) interface{} {
 		returnObj.Set("code", code)
 		returnObj.Set("cancel", js.FuncOf(func(_ js.Value, args []js.Value) interface{} {
 			cancel()
+			//return js.Global().Get("undefined")
```
Comment (bryanchriswhite) [inline wasm/client.go:247]:
```suggestion
```

**Review by bryanchriswhite — APPROVED** [review]
Thanks (again) @vu3rdd!

**Comment by vu3rdd** [comment]
> Thanks (again) @vu3rdd!
> 
> I saw a couple of lines that looked like we could delete but otherwise this looks good and works for me.

Thanks! Pushed a commit that addresses those.

**Pull request merged**